### PR TITLE
client: correct quota check in Client::_rename()

### PR DESCRIPTION
**Pull request by rishabh-d-dave**
Fixes: https://tracker.ceph.com/issues/58220

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [x] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by rishabh-d-dave** [comment]
I need to add tests for this.

**Comment by rishabh-d-dave** [comment]
workunit/quota.sh passes with this patch - http://pulpito.front.sepia.ceph.com/rishabh-2023-02-16_07:51:51-fs-wip-rishabh-fs-qa-workunit-quota.sh-testing-default-smithi/

**Comment by dparmar18** [comment]
Nice catch

I can see one more such instance in [`client/Client.cc`](https://github.com/ceph/ceph/blob/main/src/client/Client.cc#L11253)  where while setting quota root, only `QUOTA_MAX_BYTES` is considered, is this intentional?

**Comment by rishabh-d-dave** [comment]
> Nice catch
> 
> I can see one more such instance in [`client/Client.cc`](https://github.com/ceph/ceph/blob/main/src/client/Client.cc#L11253) where while setting quota root, only `QUOTA_MAX_BYTES` is considered, is this intentional?

Yes, I too noticed it. Probably this was not intentional. I am investigating into it right now.

File: src/client/Client.cc [hunk 1]
```diff
@@ -14137,9 +14137,9 @@ int Client::_rename(Inode *fromdir, const char *fromname, Inode *todir, const ch
   }
   if (cct->_conf.get_val<bool>("client_quota") && fromdir != todir) {
     Inode *fromdir_root =
-      fromdir->quota.is_enabled(QUOTA_MAX_FILES) ? fromdir : get_quota_root(fromdir, perm, QUOTA_MAX_FILES);
+      fromdir->quota.is_enabled(QUOTA_ANY) ? fromdir : get_quota_root(fromdir, perm, QUOTA_ANY);
```
Comment (vshankar) [inline src/client/Client.cc:14140]: Oh, so the quota realm for both dirs were the root realm and that allowed the rename :/

How was this test passing earlier?

**Comment by rishabh-d-dave** [comment]
jenkins test make check

File: src/client/Client.cc [hunk 2]
```diff
@@ -14137,9 +14137,9 @@ int Client::_rename(Inode *fromdir, const char *fromname, Inode *todir, const ch
   }
```
Comment (lxbsz) [inline src/client/Client.cc:14137]: @rishabh-d-dave 

Good catch. And could you add one comments here ?

`// don't allow cross-quota renames`

File: src/client/Client.cc [hunk 3]
```diff
@@ -14137,9 +14137,9 @@ int Client::_rename(Inode *fromdir, const char *fromname, Inode *todir, const ch
...
   if (cct->_conf.get_val<bool>("client_quota") && fromdir != todir) {
     Inode *fromdir_root =
-      fromdir->quota.is_enabled(QUOTA_MAX_FILES) ? fromdir : get_quota_root(fromdir, perm, QUOTA_MAX_FILES);
+      fromdir->quota.is_enabled(QUOTA_ANY) ? fromdir : get_quota_root(fromdir, perm, QUOTA_ANY);
     Inode *todir_root =
-      todir->quota.is_enabled(QUOTA_MAX_FILES) ? todir : get_quota_root(todir, perm, QUOTA_MAX_FILES);
+      todir->quota.is_enabled(QUOTA_ANY) ? todir : get_quota_root(todir, perm, QUOTA_ANY);
```
Comment (lxbsz) [inline src/client/Client.cc:14142]: You could just use `todir->quota.is_enabled()` instead to make it consistent with the existing code in other places.

File: src/client/Client.cc (hunk 1 above)
Comment (lxbsz) [reply src/client/Client.cc:14140]: This is odd.

File: src/client/Client.cc (hunk 3 above)
Comment (rishabh-d-dave) [reply src/client/Client.cc:14142]: True. `QUOTA_ANY` is the default of the only parameter for `quota.is_enabled()`. I'll make that change.

File: src/client/Client.cc (hunk 2 above)
Comment (rishabh-d-dave) [reply src/client/Client.cc:14137]: Sure.

File: src/client/Client.cc (hunk 1 above)
Comment (gregsfortytwo) [reply src/client/Client.cc:14140]: Did you check how this line got set as MAX_FILES? Was it in the original quota code, or did somebody change it deliberately?

File: src/client/Client.cc (hunk 2 above)
Comment (rishabh-d-dave) [reply src/client/Client.cc:14137]: On second thought, sounds little vague to me. I've added it but can you suggest something that offers better explanation?

File: src/client/Client.cc (hunk 1 above)
Comment (rishabh-d-dave) [reply src/client/Client.cc:14140]: Yes, without that it was hard to know if this modification is the right thing to do. The commit where this change was introduced is mentioned in my commit. In fact, that commit has one more similar change, I am working on that now.

File: src/client/Client.cc (hunk 3 above)
Comment (rishabh-d-dave) [reply src/client/Client.cc]: This change has been added.

File: src/client/Client.cc (hunk 2 above)
Comment (lxbsz) [reply src/client/Client.cc:14137]: Currently we do not allow to move a file or directory into a new path, which has a different quota realm. IMO we can support this in theory but the code maybe a bit complicated, currently both the kclient and userspace just disable cross-quota renames.

**Review by lxbsz — APPROVED** [review]
LGTM.

**Review by dparmar18 — APPROVED** [review]
LGTM

**Comment by rishabh-d-dave** [comment]
Test workunit/quota.sh now works fine with this PR - http://pulpito.front.sepia.ceph.com/rishabh-2023-02-17_02:29:36-fs-wip-rishabh-fs-qa-workunit-quota.sh-testing-default-smithi/.

**Review by vshankar — APPROVED** [review]

**Comment by rishabh-d-dave** [comment]
@vshankar Instead of applying label `wip-vshankar-backlog`,  we can apply `needs-qa` and pick PRs from there later when we are ready for QA. This way we wouldn't need more new labels and everyone could look at PRs are waiting for QA.

**Comment by vshankar** [comment]
> @vshankar Instead of applying label `wip-vshankar-backlog`,  we can apply `needs-qa` and pick PRs from there later when we are ready for QA. This way we wouldn't need more new labels and everyone could look at PRs are waiting for QA. 

needs-qa conflicts with Yuri's search, although we need to notify it for pickup. I created that label to track PRs that I need to run tests on.

**Comment by rishabh-d-dave** [comment]
> > @vshankar Instead of applying label `wip-vshankar-backlog`,  we can apply `needs-qa` and pick PRs from there later when we are ready for QA. This way we wouldn't need more new labels and everyone could look at PRs are waiting for QA.
> 
> needs-qa conflicts with Yuri's search, although we need to notify it for pickup. I created that label to track PRs that I need to run tests on.

Ah, okay. I see. Perhaps we can create separate label for Yuri called "backport-needs-qa" or something similar.

**Comment by vshankar** [comment]
> > > @vshankar Instead of applying label `wip-vshankar-backlog`,  we can apply `needs-qa` and pick PRs from there later when we are ready for QA. This way we wouldn't need more new labels and everyone could look at PRs are waiting for QA.
> > 
> > needs-qa conflicts with Yuri's search, although we need to notify it for pickup. I created that label to track PRs that I need to run tests on.
> 
> Ah, okay. I see. Perhaps we can create separate label for Yuri called "backport-needs-qa" or something similar.

I wouldn't burden Yuri with that. Better to track those ourselves.

**Comment by rishabh-d-dave** [comment]
> > > > @vshankar Instead of applying label `wip-vshankar-backlog`,  we can apply `needs-qa` and pick PRs from there later when we are ready for QA. This way we wouldn't need more new labels and everyone could look at PRs are waiting for QA.
> > > 
> > > 
> > > needs-qa conflicts with Yuri's search, although we need to notify it for pickup. I created that label to track PRs that I need to run tests on.
> > 
> > 
> > Ah, okay. I see. Perhaps we can create separate label for Yuri called "backport-needs-qa" or something similar.
> 
> I wouldn't burden Yuri with that. Better to track those ourselves.

Okay, got it.

**Comment by vshankar** [comment]
> > > > > @vshankar Instead of applying label `wip-vshankar-backlog`,  we can apply `needs-qa` and pick PRs from there later when we are ready for QA. This way we wouldn't need more new labels and everyone could look at PRs are waiting for QA.
> > > > 
> > > > needs-qa conflicts with Yuri's search, although we need to notify it for pickup. I created that label to track PRs that I need to run tests on.
> > > 
> > > Ah, okay. I see. Perhaps we can create separate label for Yuri called "backport-needs-qa" or something similar.
> > 
> > I wouldn't burden Yuri with that. Better to track those ourselves.
> 
> What? needs-qa is like a decade old and means "this PR needs test results to merge". He doesn't get to own that, especially when there are straightforward search additions like "milestone:Quincy".

These are not PRs that Yuri needs to test. It's just a tracking label that I created to pick up PRs for integration tests that I plan to run. The number of cephfs PRs were increasing since nothing was getting tested, so, I created this label for my tracking. This has absolutely nothing to do with Yuri. Where is the confusion?

**Comment by gregsfortytwo** [comment]
> These are not PRs that Yuri needs to test. It's just a tracking label that I created to pick up PRs for integration tests that I plan to run. The number of cephfs PRs were increasing since nothing was getting tested, so, I created this label for my tracking. This has absolutely nothing to do with Yuri. Where is the confusion?

You can handle your own labels if you want, just needs-qa is not his exclusive province, which you seemed to be implying. I haven't built an integration branch in a while, but if it comes up that's what I search for. It's what I set on my own PRs. Etc :)

**Comment by vshankar** [comment]
> > These are not PRs that Yuri needs to test. It's just a tracking label that I created to pick up PRs for integration tests that I plan to run. The number of cephfs PRs were increasing since nothing was getting tested, so, I created this label for my tracking. This has absolutely nothing to do with Yuri. Where is the confusion?
> 
> You can handle your own labels if you want,

That's what I meant.

> just needs-qa is not his exclusive province, which you seemed to be implying.

Nope - I didn't mean that. I just didn't want to step in each others toes with labels and search. esp. since this is a main branch PR. For backport, needs-qa is what I set for the change to show up in Yuri's search.

**Review by vshankar — APPROVED** [review]
https://tracker.ceph.com/projects/cephfs/wiki/Main#09-Mar-2023

**Pull request merged**